cpu/kinetis_common/i2c.c

```diff
@@ -0,0 +1,426 @@
+/*
+ * Copyright (C) 2014 Freie Universität Berlin
+ * Copyright (C) 2014 PHYTEC Messtechnik GmbH
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     cpu_kinetis_common_i2c
+ *
+ * @note        This driver only implements the 7-bit addressing master mode.
+ *
+ * @{
+ *
+ * @file
+ * @brief       Low-level I2C driver implementation
+ *
+ * @author      Johann Fischer <j.fischer@phytec.de>
+ *
+ * @}
+ */
+
+#include <stdint.h>
+
+#include "cpu.h"
+#include "irq.h"
+#include "mutex.h"
+#include "periph_conf.h"
+#include "periph/i2c.h"
+
+#define ENABLE_DEBUG    (0)
+#include "debug.h"
+
+/* guard file in case no I2C device is defined */
+#if I2C_NUMOF
+
+/**
+ * @brief Array holding one pre-initialized mutex for each I2C device
+ */
+static mutex_t locks[] =  {
+#if I2C_0_EN
+    [I2C_0] = MUTEX_INIT,
+#endif
+#if I2C_1_EN
+    [I2C_1] = MUTEX_INIT,
+#endif
+#if I2C_2_EN
+    [I2C_2] = MUTEX_INIT
+#endif
+#if I2C_3_EN
+    [I2C_3] = MUTEX_INIT
+#endif
+};
+
+int i2c_acquire(i2c_t dev)
+{
+    if (dev >= I2C_NUMOF) {
+        return -1;
+    }
+    mutex_lock(&locks[dev]);
+    return 0;
+}
+
+int i2c_release(i2c_t dev)
+{
+    if (dev >= I2C_NUMOF) {
+        return -1;
+    }
+    mutex_unlock(&locks[dev]);
+    return 0;
+}
+
+int i2c_init_master(i2c_t dev, i2c_speed_t speed)
+{
+    I2C_Type *i2c;
+    PORT_Type *i2c_port;
+    int pin_scl = 0;
+    int pin_sda = 0;
+
+    /* TODO: read speed configuration */
+    switch (speed) {
+        case I2C_SPEED_NORMAL:
+            break;
+
+        case I2C_SPEED_FAST:
+            break;
+
+        default:
+            return -2;
+    }
+
+    /* read static device configuration */
+    switch (dev) {
+#if I2C_0_EN
+
+        case I2C_0:
+            i2c = I2C_0_DEV;
+            i2c_port = I2C_0_PORT;
+            pin_scl = I2C_0_SCL_PIN;
+            pin_sda = I2C_0_SDA_PIN;
+            I2C_0_CLKEN();
+            I2C_0_PORT_CLKEN();
+            break;
+#endif
+
+        default:
+            return -1;
+    }
+
+    /* configure pins, alternate output */
+    i2c_port->PCR[pin_scl] = I2C_0_PORT_CFG;
+    i2c_port->PCR[pin_sda] = I2C_0_PORT_CFG;
+    /* set frequency devider register */
+    i2c->F = I2C_F_MULT(1) | I2C_F_ICR(0x1f);
```

Comment:
This will yield 100 kHz baud rate on my device (48 MHz bus clock).

Since it is a TODO to add other baud rates, add a comment with a reference to _Kinetis K60 reference manual_, section 51.4.1.10 _I2C divider and hold values_, Table 51-41. _I2C divider and hold values_ (or any other kinetis reference manual) to let us know where to look this up later after we have forgotten about it..

Add a comment that this yields a baud rate divider of 240 and a prescaler of 2.